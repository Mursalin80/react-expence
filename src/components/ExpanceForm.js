import React, { Component } from "react";

export default class ExpanceForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      note: props.expence ? props.expence.note : "",
      createdAt: props.expence ? props.expence.createdAt : 0,
      description: props.expence ? props.expence.description : "",
      amount: props.expence ? props.expence.amount : 0,
      error: ""
    };
  }

  onChangeHandler = e => {
    let { name, value } = e.target;
    this.setState(() => ({ [name]: value }));
  };

  onSubmitHandler = e => {
    e.preventDefault();
    let { note, description, amount, createdAt } = this.state;
    if (!note && !amount) {
      this.setState({ error: "No note and amount is provided." });
    }
    this.props.onSubmit({
      note,
      description,
      amount,
      createdAt
    });
  };

  render() {
    return (
      <div className="ui container">
        <form onSubmit={this.onSubmitHandler} className="ui form">
          <input
            className="field"
            type="text"
            name="note"
            placeholder="note"
            value={this.state.note}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="amount"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="description"
            placeholder="description"
            value={this.state.description}
            onChange={this.onChangeHandler}
          />
          <input
            className="field"
            type="text"
            name="createdAt"
            placeholder="createdAt"
            value={this.state.createdAt}
            onChange={this.onChangeHandler}
          />

          <button className="ui button">Add Expence</button>
        </form>
      </div>
    );
  }
}
