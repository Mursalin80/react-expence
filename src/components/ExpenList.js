import React from "react";
import { connect } from "react-redux";
import ExpenListItems from "./ExpenListItems";
// import { store } from "../redux/store";

export const ExpenList = props => {
  let { expences } = props;
  return (
    <div>
      {expences.length === 0 ? (
        <p>No Expences</p>
      ) : (
        expences.map(({ createdAt, note, amount, id }) => (
          <ExpenListItems
            key={id}
            createdAt={createdAt}
            note={note}
            amount={amount}
            id={id}
          />
        ))
      )}
    </div>
  );
};

const mapStateToProps = state => ({ expences: state.expenReducer });

export default connect(mapStateToProps)(ExpenList);
