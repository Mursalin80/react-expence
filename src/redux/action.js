import { v4 as uuidv4 } from "uuid";

export const getVisibalExpances = (expances, { note = "" } = {}) => {
  return expances.filter(exp => exp.note.includes(note));
};

export const getVisExpSortBy = (
  expances,
  { note = "", sortBy = "date" } = {}
) => {
  return expances
    .filter(exp => exp.note.includes(note))
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt > b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? 1 : -1;
      }
    });
};

// add expance action
export const addExpance = ({
  note = "",
  amount = 0,
  desc = "",
  createdAt = 0
} = {}) => ({
  type: "Add_exp",
  expence: { id: uuidv4(), note, amount, desc, createdAt }
});

// remove expance action
export const removeExpance = ({ id } = {}) => ({
  type: "Remove_exp",
  id
});

// Edit expance action
export const editExpance = ({ id, update }) => ({
  type: "Edit_exp",
  id,
  payload: update
});

// add note filter action

export const noteFilter = (note = "") => ({
  type: "Note_filter",
  note
});

// sortBy filter action

export const sortByDate = () => ({
  type: "Sort_by_Date"
});

export const sortByAmount = () => ({
  type: "Sort_by_Amount"
});
