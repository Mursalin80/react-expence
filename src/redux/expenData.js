export const expData = [
  {
    id: 1,
    note: "gas bill",
    description: "gas bill for month of Jan 2020",
    amount: 250,
    createdAt: 2000
  },
  {
    id: 2,
    note: "net bill",
    description: "net bill for month of Jan 2020",
    amount: 1000,
    createdAt: 3600
  },
  {
    id: 3,
    note: "Milk bill",
    description: "Milk bill for month of Jan 2020",
    amount: 5600,
    createdAt: 5300
  },
  {
    id: 4,
    note: "Milk bill feb",
    description: "Milk bill for month of Feb 2020",
    amount: 5809,
    createdAt: 6000
  },
  {
    id: 5,
    note: "Milk bill mar",
    description: "Milk bill for month of Mar 2020",
    amount: 4500,
    createdAt: 6500
  }
];
