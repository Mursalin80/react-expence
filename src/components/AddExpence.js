import React from "react";
import ExpanceForm from "./ExpanceForm";
import { connect } from "react-redux";
import { addExpance } from "../redux/action";

export const AddExpence = ({ onSubmit }) => {
  return (
    <div>
      <h1>Add Expence</h1>
      <ExpanceForm
        onSubmit={expence => {
          onSubmit(expence);
        }}
      />
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: expence => dispatch(addExpance(expence))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpence);
