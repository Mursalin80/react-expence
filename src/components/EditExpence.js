import { connect } from "react-redux";
import ExpanceForm from "./ExpanceForm";
import { editExpance } from "../redux/action";

import React, { Component } from "react";

export class EditExpence extends Component {
  onSubmit = expence => {
    this.props.editExpance(this.props.expence.id, expence);
  };
  render() {
    return (
      <div>
        <ExpanceForm expence={this.props.expence} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateTopProps = (state, props) => {
  return {
    expence: state.expenReducer.find(exp => exp.id === props.match.params.id)
  };
};

const mapDispatchToPrors = dispatch => {
  return {
    editExpance: (id, expence) => dispatch(editExpance(id, expence))
  };
};

export default connect(mapStateTopProps, mapDispatchToPrors)(EditExpence);
