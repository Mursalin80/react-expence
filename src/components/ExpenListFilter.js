import React from "react";
import { connect } from "react-redux";
import { noteFilter } from "../redux/action";

const ExpenListFilter = ({ filter, dispacth }) => {
  return (
    <div className="ui form">
      <input
        className="ui field"
        type="text"
        value={filter.note}
        onChange={e => {
          dispacth(noteFilter(e.target.value));
        }}
      />
      <select
        className="ui fluid dropdown"
        value={filter.sortBy}
        onChange={e => {
          if (e.target.value === "date") {
          } else if (e.target.value === "amount") {
          }
        }}
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>
    </div>
  );
};

const mapStateToProps = state => {
  return { filter: state.filterReducer };
};

export default connect(mapStateToProps)(ExpenListFilter);
