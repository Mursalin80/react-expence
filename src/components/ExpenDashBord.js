import React from "react";
import ExpenList from "./ExpenList";
import ExpenListFilter from "./ExpenListFilter";
import AddExpence from "./AddExpence";

export default function ExpenDashBord() {
  return (
    <div>
      <h1>Dash Bord</h1>
      <AddExpence />
      <ExpenListFilter />
      <ExpenList />
    </div>
  );
}
