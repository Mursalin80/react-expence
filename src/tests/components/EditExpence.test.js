import { shallow } from "enzyme";
import React from "react";
import toJson from "enzyme-to-json";
import "../Enyzme.config";
import { EditExpence } from "../../components/EditExpence";
import { expData } from "../../redux/expenData";
import { editExpance } from "../../redux/action";

let editExp, wrap;

beforeEach(() => {
  editExp = jest.fn();
  wrap = shallow(<EditExpence editExpance={editExp} expence={expData[1]} />);
});

test("should render add expence", () => {
  expect(toJson(wrap)).toMatchSnapshot();
});
