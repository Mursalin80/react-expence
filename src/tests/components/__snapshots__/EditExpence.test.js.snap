// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expence 1`] = `
<div>
  <ExpanceForm
    expence={
      Object {
        "amount": 1000,
        "createdAt": 3600,
        "description": "net bill for month of Jan 2020",
        "id": 2,
        "note": "net bill",
      }
    }
    onSubmit={[Function]}
  />
</div>
`;
